pkg/protocol/chain.go

```diff
@@ -0,0 +1,400 @@
+// latest produced commitment).
+func (c *Chain) deriveVerifiedWeight() func() {
+	return c.VerifiedWeight.DeriveValueFrom(reactive.NewDerivedVariable(func(_ uint64, latestProducedCommitment *Commitment) uint64 {
+		if latestProducedCommitment == nil {
+			return 0
+		}
+
+		return latestProducedCommitment.CumulativeWeight()
+	}, c.LatestProducedCommitment))
+}
+
+// deriveChildChains defines how a chain determines its ChildChains (by adding each child to the set).
+func (c *Chain) deriveChildChains(child *Chain) func() {
+	c.ChildChains.Add(child)
+
+	return func() {
+		c.ChildChains.Delete(child)
+	}
+}
+
+// deriveParentChain defines how a chain determines its parent chain from its forking point (it inherits the Chain from
+// the parent commitment of the forking point or nil if either of them is still unknown).
+func (c *Chain) deriveParentChain(forkingPoint *Commitment) (shutdown func()) {
+	if forkingPoint != nil {
+		return forkingPoint.Parent.WithValue(func(parentCommitment *Commitment) (shutdown func()) {
+			if parentCommitment != nil {
+				return c.ParentChain.InheritFrom(parentCommitment.Chain)
+			}
+
+			c.ParentChain.Set(nil)
+
+			return nil
+		})
+	}
+
+	c.ParentChain.Set(nil)
+
+	return nil
+}
+
+// deriveOutOfSyncThreshold defines how a chain determines its "out of sync" threshold (the latest seen slot minus 2
+// times the max committable age or 0 if this would cause an overflow to the negative numbers).
+func (c *Chain) deriveOutOfSyncThreshold(latestSeenSlot reactive.ReadableVariable[iotago.SlotIndex], engineInstance *engine.Engine) func() {
+	return c.OutOfSyncThreshold.DeriveValueFrom(reactive.NewDerivedVariable(func(_ iotago.SlotIndex, latestSeenSlot iotago.SlotIndex) iotago.SlotIndex {
+		if outOfSyncOffset := 2 * engineInstance.LatestAPI().ProtocolParameters().MaxCommittableAge(); outOfSyncOffset < latestSeenSlot {
+			return latestSeenSlot - outOfSyncOffset
+		}
+
+		return 0
+	}, latestSeenSlot))
+}
+
+// deriveWarpSyncThreshold defines how a chain determines its warp sync threshold (the latest seen slot minus the max
+// committable age or 0 if this would cause an overflow to the negative numbers).
+func (c *Chain) deriveWarpSyncThreshold(latestSeenSlot reactive.ReadableVariable[iotago.SlotIndex], engineInstance *engine.Engine) func() {
+	return c.WarpSyncThreshold.DeriveValueFrom(reactive.NewDerivedVariable(func(_ iotago.SlotIndex, latestSeenSlot iotago.SlotIndex) iotago.SlotIndex {
+		if warpSyncOffset := engineInstance.LatestAPI().ProtocolParameters().MaxCommittableAge(); warpSyncOffset < latestSeenSlot {
+			return latestSeenSlot - warpSyncOffset
+		}
+
+		return 0
+	}, latestSeenSlot))
+}
+
+// addCommitment adds the given commitment to this chain.
+func (c *Chain) addCommitment(newCommitment *Commitment) (shutdown func()) {
+	c.commitments.Set(newCommitment.Slot(), newCommitment)
+
+	c.LatestCommitment.Set(newCommitment)
```

Comment:
How do we make sure that this commitment is really the latest for this chain?

Reply:
commitments are associated with a chain in solidification order, so they arrive here in the correct order,